src/server/v2.0/handler/user.go

```diff
@@ -0,0 +1,485 @@
+	"github.com/goharbor/harbor/src/controller/user"
+	"github.com/goharbor/harbor/src/jobservice/logger"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/permission/types"
+	usermodels "github.com/goharbor/harbor/src/pkg/user/models"
+	"github.com/goharbor/harbor/src/server/v2.0/handler/model"
+	"github.com/goharbor/harbor/src/server/v2.0/models"
+	"github.com/goharbor/harbor/src/server/v2.0/restapi"
+	"github.com/goharbor/harbor/src/server/v2.0/restapi/operations/users"
+)
+
+var userResource = system.NewNamespace().Resource(rbac.ResourceUser)
+
+type usersAPI struct {
+	BaseAPI
+	ctl     user.Controller
+	getAuth func(ctx context.Context) (string, error) // For testing
+}
+
+func (u *usersAPI) SetCliSecret(ctx context.Context, params users.SetCliSecretParams) middleware.Responder {
+	uid := int(params.UserID)
+	if err := u.requireForCLISecret(ctx, uid); err != nil {
+		return u.SendError(ctx, err)
+	}
+	if err := requireValidSecret(params.Secret.Secret); err != nil {
+		return u.SendError(ctx, err)
+	}
+	if err := u.ctl.SetCliSecret(ctx, uid, params.Secret.Secret); err != nil {
+		logger.Errorf("Failed to set CLI secret, error: %v", err)
+		return u.SendError(ctx, err)
+	}
+	return users.NewSetCliSecretOK()
+}
+
+func (u *usersAPI) CreateUser(ctx context.Context, params users.CreateUserParams) middleware.Responder {
+	logger := log.GetLogger(ctx)
+	if err := u.requireCreatable(ctx); err != nil {
+		return u.SendError(ctx, err)
+	}
+	if err := requireValidSecret(params.UserReq.Password); err != nil {
+		return u.SendError(ctx, err)
+	}
+	m := &usermodels.User{
+		Username: params.UserReq.Username,
+		Realname: params.UserReq.Realname,
+		Email:    params.UserReq.Email,
+		Comment:  params.UserReq.Comment,
+		Password: params.UserReq.Password,
+	}
+	if err := validateUserProfile(m); err != nil {
+		return u.SendError(ctx, err)
+	}
+	uid, err := u.ctl.Create(ctx, m)
+	if err != nil {
+		logger.Errorf("Failed to create user, error: %v", err)
+		return u.SendError(ctx, err)
+	}
+	location := fmt.Sprintf("%s/%d", strings.TrimSuffix(params.HTTPRequest.URL.Path, "/"), uid)
+	return users.NewCreateUserCreated().WithLocation(location)
+
+}
+
+func (u *usersAPI) ListUsers(ctx context.Context, params users.ListUsersParams) middleware.Responder {
+	if err := u.RequireSystemAccess(ctx, rbac.ActionList, userResource); err != nil {
+		return u.SendError(ctx, err)
+	}
+	query, err := u.BuildQuery(ctx, params.Q, params.Sort, params.Page, params.PageSize)
+	if err != nil {
+		return u.SendError(ctx, err)
+	}
+	values := params.HTTPRequest.URL.Query()
+	for _, k := range []string{"username", "email"} {
+		if v := values.Get(k); v != "" {
+			query.Keywords[k] = &q.FuzzyMatchValue{Value: v}
+		}
+	}
+	total, err := u.ctl.Count(ctx, query)
+	if err != nil {
+		return u.SendError(ctx, err)
+	}
```

Comment:
May return directly here when the total is zero to reduce a DB query.

Reply:
good point, fixed.